README.md
[![Build Status](https://travis-ci.org/braydenbekker/EAM-MBTR.svg?branch=master)](https://travis-ci.org/braydenbekker/EAM-MBTR)
[![codecov](https://codecov.io/gh/braydenbekker/EAM-MBTR/branch/master/graph/badge.svg)](https://codecov.io/gh/braydenbekker/EAM-MBTR)
## My Research Impact


Provide and extensive proof of concept for MBTR, MPT, and High-Thoughput methods
and there ability to make fast and accurate out-of sample predictions for binary, ternary and quaternary system.

#### abstract
-------------

One of the central problems in materials research is accurately predicting the properties of an alloy. Computational materials discovery has the potential to open the floodgates for human progression. Our society, built upon the materials discoveries of the past, lies in wait for the discovery of the next material that will change the world. 
\end{abstract}

#### Introduction
-----------------

In the past metallurgists made there discoveries through an exhaustive trial and error process which eventually resulted in a happy mistake. 

<img src="/ProgramFiles/RoyPlunkett.png" width="400">

Roy Plunkett was trying to create a new refrigerant when a buildup of a waxy material started to form on the inside of the container of CFC gasses he was working with. Teflon soon followed.
With computers came computational methods which remove much of the guess work, however, these modern processes of materials prediction are still too slow. 

<img src="/ProgramFiles/materialscooking.png" width="400">

Imagine for me a stocked pantry, to create a dessert by trial and error would be a time consuming disaster. A cookbook provides necessary directions. If you know the recipe than the desired result can be achieved. If you imagine now, a materials pantry stocked with every possible configuration of elements the number of combinations are infinite. The existing computational methods like DFT (Density Functional Theory) act as materials cookbooks. While the existing "materials cookbooks" remove the guess work they have a time consuming method for searching the "table of contents" to continue or example. DFT makes accurate predictions but at high computational cost.    


<img src="/ProgramFiles/MBTR.png" width="400">

My current research is with the recently proposed Many Body Tensor Representation (MBTR) for atomic systems. This machine learning approach uses the existing materials "cookbooks" to train and then make out-of-sample predictions. My project researches the ability of MBTR to replace the existing materials "cookbooks" using machine learning to predict new materials at a faster rate.

#### Main Objective
-------------------

- Search the Interatomic Potentials Repository for binary/ternary systems with EAM potentials

- Use Enumlib to get the derivative superstructures for each system

- For each system from FCC, BCC, and HCP choose 2500 derivative structures: the first 500 and 2000 additional structures from 10-12 atom cells.  

- Relax the atomic positions and compute energies per atom and enthalpy.

- Plot the convex hull and verify with AFLOW that results are similar (EAM potentials will not match results, here we are only looking for general similarities.)

- Save the data for each system to a HDF5 file for ease of data access.

- Use results to train and predict in MBTR

Full API Documentation available at: [github pages](https://github.com/braydenbekker/braydenbekker.github.io/blob/master/eamEnergies/html/index.html). 


HISTORY.md
# Revision History for eamEnergies

## Revision 0.0.1
- Initial code commit of getStructures.py.
- Updated README.md to have API Documentation link, and badges for
  build status, code coverage, and code quality.
- Added files for Travis CI.
- Added files to create python package.

tests/test_demo.py
"""Tests the mathematical functions defined in eamEnergies/getStructures.py
"""

import pytest

def test_GenEnumOut():
    """Tests the squaring function"""
    from eamEnergies.getStructures import GenEnumOut, GenVaspFiles

    lattice = ["fcc","bcc","hcp"]
    
    arg = ["filename", "Binary"]
    assert 1 == GenEnumOut(lattice, arg)
    
    arg = ["filename", "Ternary"]
    assert 1 == GenEnumOut(lattice, arg)

    arg = ["filename", "Al", "Ti"]
    assert 1 == GenVaspFiles(lattice, arg, test=True, conc1="100 200 5", conc2="1 5")

    arg = ["filename", "Al", "Co", "Ni"]
    assert 1 == GenVaspFiles(lattice, arg, test=True, conc1="100 200 5", conc2="1 5")
    return



eamEnergies/trial.py
'''
Author Brayden Bekker

getStructures.py uses the enumlib code created by the MSG group at BYU, https://github.com/msg-byu/enumlib. enumlib is 
designed to compute the derivative superstructures for a system of elements and a given lattice. 

Input:
    - path to makeStr.py and enum.x from enumlib.
    - precomputed struct_enum.out files for FCC, BCC, and HCP through 7140 structures.

Output:
    - a directory (e.g. Structures_AlTi) whith sub directories, fcc, bcc, hcp
    - each subdirectory contains 2500 vasp POSCAR files with atom concentrations and positions.
    (the first 500 derivative superstructures and 2000 additional structures from higher atom cells 10-12)
    
File Structure:
cd path/to/program/files/
ls
Structures/ struct_enum.in.fcc, struct_enum.in.bcc, struct_enum.in.hcp
enum_(fcc,bcc,hcp)/struct_enum.out
makeStr.py
./enum.x
'''
import shutil
import numpy as np
import os
import random
import sys

def square(x):
    """Finds the square of the input.
    
    Args:
        x (float): The number to be squared.

    Returns:
        x2 (float): The squared number.
    """

    return x**2


eamEnergies/getStructures.py
'''
Author Brayden Bekker

getStructures.py uses the enumlib code created by the MSG group at BYU, https://github.com/msg-byu/enumlib. enumlib is 
designed to compute the derivative superstructures for a system of elements and a given lattice. 

Input:
    - path to makeStr.py and enum.x from enumlib.
    - precomputed struct_enum.out files for FCC, BCC, and HCP through 7140 structures.

Output:
    - a directory (e.g. Structures_AlTi) whith sub directories, fcc, bcc, hcp
    - each subdirectory contains 2500 vasp POSCAR files with atom concentrations and positions.
    (the first 500 derivative superstructures and 2000 additional structures from higher atom cells 10-12)
    
Notes:
    lines with #pragma: no cover have been extensively unit tested through commandline tests and were exculded here
    as they could not be unit tested by pytest without parsing argv and adding values at run time.
'''
import shutil
import numpy as np
import os
import random
import sys

def GenEnumOut(structure, arg):
    '''
    Compute the struct_enum.out vasp file.

    Args:
        enum.x (fortran, enumlib): computes derivative superstructures of a system.
        struct_enum.in (vasp, INCAR): vasp input file for each lattice structure
    Returns:
        enum_'lattice'/'Binary or Ternary'/Struct_enum.out
    '''
    if(len(sys.argv)>1): arg = sys.argv 

    if(len(arg[1])>2): 
        for lattice in structure:
            os.chdir("enum_%s/%s/"%(lattice, arg[1]))

            input_file=('struct_enum.in')
        
            enum="./../../enum.x %s" %(input_file)
            os.system(enum)
            os.chdir("../../")
        return 1
    return 0 #pragma: no cover

def GenVaspFiles(structure, arg, test,  conc1="1211 7140 200", conc2="1 500"):
    '''
    Generate 7500 vasp files, 2500 from each chosen lattice type.

    Args:
        structure (np.array): array of chosen lattice types
        struct_enum.out (vasp. OUTCAR): precomputed vasp outcar for each lattice type
        system (argv): 2-4 space separated elements as command line args
    Returns:
        Structure_'system'/'fcc/, bcc/, hcp'/vasp.num where num=1-2500
    '''
    if(len(sys.argv)>1): arg = sys.argv

    elements = []
    system = ""
    directory="Structures_"
    for i in arg:
        elements.append(i) #get the selected elements from command line                                        
        if(len(i)==2): 
            directory=(directory + "%s"%(i))
            system = system + " " + i
    if(len(system)==6): systemType="Binary"
    elif(len(system)==9): systemType="Ternary"
    
    try:
        os.mkdir(directory)
    except OSError: #pragma: no cover
        os.system("rm -rf %s"%(directory))
        os.mkdir(directory)
    
    homedirectory = os.getcwd()
    workingdirectory = ("%s/%s"%(os.getcwd(), directory)) #file path to enumlib files             
    os.chdir(workingdirectory)


    for lattice in structure:
        directory_lattice="%s"%(lattice)
        os.mkdir(directory_lattice)
        os.chdir("./" + directory_lattice + "/")
        os.system(("cp %s/enum_%s/%s/struct_enum.out %s") 
                  %(homedirectory, lattice, systemType, os.getcwd()))
    
        makestructures="python %s/makeStr.py %s -species %s" %(homedirectory, conc1, system)
        os.system(makestructures)

        makestructures="python %s/makeStr.py %s -species %s" %(homedirectory, conc2, system)
        os.system(makestructures)

        os.system("rm struct_enum.out")
        os.chdir("../")

    return 1

def main():
    '''
    enumlib computes the derivative superstructures of a system. 
    getStructures.py has two principle classes: compute the struct_enum.out using enum.x,
    second, compute the derivative superstructures using the struct_enum.out for fcc, bcc, hcp.

    Args:
         struct_enum.in.'struct' (vasp POSCAR): binary, ternary, quaternary POSCAC file
         enum.x (fortran, enumlib): required code for creating struct_enum.out
         makeStr.py (Python, enumlib): edited from oririnal to allow for random sampling
         system (argv, 2-4): 2 to 4 element names, space separated (e.g. Al Ti) 

    Returns: 
         struct_enum.out (vasp OUTCAR): the outcar must contain >7140 
                       structures that is up to 12 atom cells for fcc
         Structure_'system'/fcc /bcc /hcp /vasp.'num' (vasp files): num=(1-2500), first 1-500 structures,
                       second 500-2500 sampled  randomly for 10-12 atom cells or more generally between 1211-7140  
    '''
    structure = np.array(["fcc","bcc","hcp"]) #the three desired crystal lattice types 
    if(len(sys.argv) > 1):
        ran =  GenEnumOut(structure, arg=[""]) #pragma: no cover

        if(ran == 0): GenVaspFiles(structure, arg=[""], test=False, conc1="100 200 5", conc2="1 5") #pragma: no cover
    return 0

main()
